caffe2/core/c10_operator.h

```diff
@@ -8,11 +8,22 @@
 namespace caffe2 {
 namespace detail {
 
-using _CallCaffe2OpFunc = void(const c10::FunctionSchema& schema, std::vector<c10::IValue>& inputs, std::vector<c10::IValue*>& outputs);
+constexpr const char* PREALLOCATED_OUTPUT_ARGNAME =
+    "_caffe2_preallocated_outputs";
 
-template<class Caffe2Operator>
-inline void _call_caffe2_op(const c10::FunctionSchema& schema, std::vector<c10::IValue>& inputs, std::vector<c10::IValue*>& outputs) {
-  Caffe2Operator(schema, std::move(inputs), std::move(outputs)).Run();
+using _CallCaffe2OpFunc = std::vector<at::Tensor>(
+    const c10::FunctionSchema& schema,
+    std::vector<c10::IValue>&& inputs,
+    std::vector<at::Tensor>&& outputs);
```

Comment:
Can you remind me why we can't use a stack here?

Reply:
In particular, is there any reason to specifically pass a freshly allocated vector of `c10::IValue` into the Caffe2Op. It seems like you can just pass in the stack and let the operator handle the popping and pushing.

Reply:
The issue is that caffe2 operators have to take inputs/outputs in the constructor and not when called (because many operator constructors actually access them to read nontensor arguments).
We could pass in a stack and the op could read inputs from it, but it would have no way to return outputs through it. I have a few ideas and will experiment with it but it'd be a larger refactoring and I'm not going to do it on this diff.